ethcore/verification/src/verification.rs

```diff
@@ -332,12 +321,36 @@ fn verify_parent(header: &Header, parent: &Header, engine: &dyn Engine) -> Resul
 	if header.number() == 0 {
 		return Err(BlockError::RidiculousNumber(OutOfBounds { min: Some(1), max: None, found: header.number() }).into());
 	}
+	if let Some(gas_limit) = engine.gas_limit_override(header) {
+		if *header.gas_limit() != gas_limit {
+			return Err(BlockError::InvalidGasLimit(
+				OutOfBounds { min: Some(gas_limit), max: Some(gas_limit), found: *header.gas_limit() }
+			).into());
+		}
+	} else {
+		let min_gas_limit = engine.min_gas_limit();
+		if header.gas_limit() < &min_gas_limit {
+			return Err(BlockError::InvalidGasLimit(
+				OutOfBounds { min: Some(min_gas_limit), max: None, found: *header.gas_limit() }
+			).into());
+		}
+		if let Some(limit) = engine.maximum_gas_limit() {
+			if header.gas_limit() > &limit {
+				return Err(BlockError::InvalidGasLimit(
+					OutOfBounds { min: None, max: Some(limit), found: *header.gas_limit() }
+				).into());
+			}
+		}
```

Comment:
this function is called `verify_parent`, but this logic has nothing to do with the parent header

Reply:
what if you move this verification to engine's `fn verify_block_basic`? If you do that you will not need to add additional methods to the `Engine` trait

Reply:
Done.
I also moved the checks back to `verify_header_params` that were originally there.
Since I'm now calling `gas_limit_override` three times instead of once, I added a cache to memoize the results, to avoid executing the same contract call multiple times.

Reply:
Thanks for changes, but that's not exactly what I wanted 😅 

I believe that files:
- `ethcore/verification/src/verification.rs`
- `ethcore/engine/src/engine.rs`

can and should not be modified by this pr.

All the changes can be encapsulated in `ethcore/engines/authority-round/src/lib.rs` file

Reply:
How could this be achieved without modifying `ethcore/verification/src/verification.rs`? After all, it's an exception to block verification to allow for gas limits set by the engine.

Reply:
What we could do is move all the gas limit related verification code from `verification.rs` to the engines, and make the current behavior the `Engine` trait's default implementation. Do you think that would be better?